File: medium/3453/main.py
class Solution:
    def separateSquares(self, squares: List[List[int]]) -> float:
        totalArea = 0
        height = float('-inf')
        low = 0
        ansY = 0
        for i in range(len(squares)):
            totalArea += squares[i][2] * squares[i][2]
            height = max(height, squares[i][1] + squares[i][2])

        while (height - low > 1e-5):
            ansY = (height + low)/2
            aboveArea = 0
            for i in range(len(squares)):
                if squares[i][1] > ansY:
                    aboveArea += squares[i][2] * squares[i][2]
                elif squares[i][1] + squares[i][2] > ansY:
                    aboveArea += (squares[i][1] + squares[i][2] - ansY) * squares[i][2]

            if aboveArea * 2 > totalArea:
                low = ansY
            else:
                height = ansY

        return ansY


        

File: medium/3453/main.c
#include <limits.h>
#include <math.h>

#define PERCISION 0.00001

double separateSquares(int** squares, int squaresSize, int* squaresColSize) {
  double totalArea = 0;
  double height = INT_MIN;
  double low = 0.0;
  double ansY;

  for (int i = 0; i < squaresSize; i++) {
    totalArea += (double)squares[i][2] * (double)squares[i][2];
    height = fmax(height, squares[i][1] + squares[i][2]);
  }

  while (height - low > PERCISION) {
    ansY = (height + low) / 2.0;
    double aboveArea = 0.0;
    for (int i = 0; i < squaresSize; i++) {
      if (squares[i][1] > ansY) {
        aboveArea += (double)squares[i][2] * (double)squares[i][2];
      } else if (squares[i][1] + squares[i][2] > ansY) {
        aboveArea += (double)(squares[i][1] + squares[i][2] - ansY) * (double)squares[i][2];
      }
    }
    // If aboveArea is more than half of totalArea, we need to increase low bound.
    // The division is more cycle efficient than multiplying.
    if (aboveArea * 2 > totalArea) {
      low = ansY;
    } else {
      height = ansY;
    }
  }

  return ansY;
    
}
